go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -2628,6 +2631,294 @@ func (s *VtctldServer) ValidateShard(ctx context.Context, req *vtctldatapb.Valid
+	}
+
+	if req.IncludeVschema {
+		validateVSchemaReq := &vtctldatapb.ValidateVSchemaRequest{
+			Keyspace:      req.Keyspace,
+			Shards:        []string{req.Shard},
+			ExcludeTables: req.ExcludeTables,
+			IncludeViews:  req.IncludeViews,
+		}
+
+		if _, err := s.ValidateVSchema(ctx, validateVSchemaReq); err != nil {
+			return nil, vterrors.Wrapf(err, "ValidateVSchema(%v) failed: %v", validateVSchemaReq, err)
+		}
+	}
+
+	// read all the aliases in the shard, that is all tablets that are
+	// replicating from the primary
+	aliases, err := s.ts.FindAllTabletAliasesInShard(ctx, req.Keyspace, req.Shard)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "FindAllTabletAliasesInShard(%v, %v) failed: %v", req.Keyspace, req.Shard, err)
+	}
+
+	var (
+		wg  sync.WaitGroup
+		rec concurrency.AllErrorRecorder
+	)
+
+	for _, alias := range aliases {
+		if topoproto.TabletAliasEqual(alias, si.Shard.PrimaryAlias) {
+			continue
+		}
+
+		wg.Add(1)
+		go func(alias *topodatapb.TabletAlias) {
+			defer wg.Done()
+			if err := s.diffSchema(ctx, primarySchema, si.Shard.PrimaryAlias, alias, req.ExcludeTables, req.IncludeViews); err != nil {
+				rec.RecordError(err)
+			}
+		}(alias)
+	}
+
+	wg.Wait()
+
+	if rec.HasErrors() {
+		return nil, vterrors.Wrapf(rec.Error(), "schema diffs: %v", rec.Error().Error())
+	}
+	return &vtctldatapb.ValidateSchemaShardResponse{}, nil
+}
+
+// ValidateVSchema is part of the vtctlservicepb.VtctldServer interface
+func (s *VtctldServer) ValidateVSchema(ctx context.Context, req *vtctldatapb.ValidateVSchemaRequest) (*vtctldatapb.ValidateVSchemaResponse, error) {
+	span, ctx := trace.NewSpan(ctx, "VtctldServer.ValidateVSchema")
+	defer span.Finish()
+
+	span.Annotate("keyspace", req.Keyspace)
+	span.Annotate("shards", strings.Join(req.Shards, ","))
+	span.Annotate("include_views", req.IncludeViews)
+	span.Annotate("exclude_tables", strings.Join(req.ExcludeTables, ","))
+
+	vschm, err := s.ts.GetVSchema(ctx, req.Keyspace)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetVSchema(%s) failed: %v", req.Keyspace, err)
+	}
+
+	var (
+		wg            sync.WaitGroup
+		shardFailures concurrency.AllErrorRecorder
+	)
+
+	for _, shard := range req.Shards {
+		wg.Add(1)
+		go func(shard string) {
+			defer wg.Done()
+			notFoundTables := []string{}
+			si, err := s.ts.GetShard(ctx, req.Keyspace, shard)
+			if err != nil {
+				shardFailures.RecordError(vterrors.Wrapf(err, "GetShard(%v, %v) failed: %v", req.Keyspace, shard, err))
+				return
+			}
+
+			primaryTablet, err := s.ts.GetTablet(ctx, si.Shard.PrimaryAlias)
+			if err != nil {
+				shardFailures.RecordError(vterrors.Wrapf(err, "GetTablet(%v) failed: %v", topoproto.TabletAliasString(si.Shard.PrimaryAlias), err))
+				return
+			}
+
+			primarySchema, err := s.tmc.GetSchema(ctx, primaryTablet.Tablet, nil, req.ExcludeTables, req.IncludeViews)
```

Comment:
`s.GetSchema(...)`

Reply:
note: in this case, I think you can get away with `TableNamesOnly: true` in the request, which will noticeably speed things up, especially for tablets with many tables or tables with large schemas